src/Database/LSMTree/Internal.hs

```diff
@@ -374,39 +374,47 @@ withSession tr hfs hbio dir = bracket (openSession tr hfs hbio dir) closeSession
 -- | See 'Database.LSMTree.Common.openSession'.
 openSession ::
      forall m h.
-     (MonadCatch m, MonadSTM m, MonadMVar m)
+     (MonadSTM m, MonadMVar m, PrimMonad m, MonadMask m)
   => Tracer m LSMTreeTrace
   -> HasFS m h
   -> HasBlockIO m h -- TODO: could we prevent the user from having to pass this in?
   -> FsPath -- ^ Path to the session directory
   -> m (Session m h)
-openSession tr hfs hbio dir = do
-    traceWith tr (TraceOpenSession dir)
-    dirExists <- FS.doesDirectoryExist hfs dir
-    unless dirExists $
-      throwIO (SessionDirDoesNotExist (FS.mkFsErrorPath hfs dir))
-    -- List directory contents /before/ trying to acquire a file lock, so that
-    -- that the lock file does not show up in the listed contents.
-    dirContents <- FS.listDirectory hfs dir
-    -- Try to acquire the session file lock as soon as possible to reduce the
-    -- risk of race conditions.
-    --
-    -- The lock is only released when an exception is raised, otherwise the lock
-    -- is included in the returned Session.
-    bracketOnError
-      acquireLock
-      releaseLock
-      $ \case
-          Left e
-            | FS.FsResourceAlreadyInUse <- FS.fsErrorType e
-            , fsep@(FsErrorPath _ fsp) <- FS.fsErrorPath e
-            , fsp == lockFilePath
-            -> throwIO (SessionDirLocked fsep)
-          Left  e -> throwIO e -- rethrow unexpected errors
-          Right Nothing -> throwIO (SessionDirLocked (FS.mkFsErrorPath hfs lockFilePath))
-          Right (Just sessionFileLock) ->
-            if Set.null dirContents then newSession sessionFileLock
-                                    else restoreSession sessionFileLock
+openSession tr hfs hbio dir =
+    -- We can not use modifyWithActionRegistry here, since there is no in-memory
+    -- state to modify. We use withActionRegistry instead, which may have a tiny
+    -- chance of leaking resources if openSession is not called in a masked
+    -- state.
+    withActionRegistry $ \reg -> do
+      traceWith tr (TraceOpenSession dir)
+      dirExists <- FS.doesDirectoryExist hfs dir
+      unless dirExists $
+        throwIO (SessionDirDoesNotExist (FS.mkFsErrorPath hfs dir))
+      -- List directory contents /before/ trying to acquire a file lock, so that
+      -- that the lock file does not show up in the listed contents.
+      dirContents <- FS.listDirectory hfs dir
+      -- Try to acquire the session file lock as soon as possible to reduce the
+      -- risk of race conditions.
+      --
+      -- The lock is only released when an exception is raised, otherwise the lock
+      -- is included in the returned Session.
+      elock <-
+        withRollbackFun reg
+          (\case Right x -> x; _ -> Nothing)
```

Comment:
Hmm I suppose this is a roundabout way of writing `fromRight Nothing` ?

Reply:
Or using just Prelude functions: `either Nothing id`